include/uk/event: Make handler and raise return values type-safe

michpappas opened the pull request:
`uk_event` handlers often omit using one of the designated macro values, and return an integer instead. This is particularly problematic when returning zero to signify that the event was handled, as `event.h` assigns zero to `UK_EVENT_NOT_HANDLED`.

Convert `uk_event` handler return values to enum to catch such errors at compile-time. Introduce a `UK_EVENT_ERROR` state and an additional parameter in the handler prototype that handlers must populate before returning `UK_EVENT_ERROR`.

<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

michpappas committed:
include/uk/event: Make handler and raise return values type safe

uk_event handlers often omit using one of the designated macro values,
and return an integer instead. This is particularly problematic when
returning zero to signify that the event was handled, as event.h
assigns zero to UK_EVENT_NOT_HANDLED.

Convert uk_event handler return values to enum to catch such errors
at compile-time. Introduce a UK_EVENT_ERROR state and an additional
parameter in the handler prototype that handlers must populate before
returning UK_EVENT_ERROR.

Checkpatch-ignore: MACRO_ARG_REUSE
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-fdtab: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukdebug: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-socket: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Checkpatch-ignore: MACRO_ARG_REUSE
Checkpatch-ignore: MACRO_ARG_PRECEDENCE
Checkpatch-ignore: UNNECESSARY_PARENTHESES
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukvmem: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukboot: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/syscall-shim: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/uknofault: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Checkpatch-Ignore: LONG_LINE
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukstore: Adapt to strongly typed uk_event API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
plat/common: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on include/uk/event.h:
```diff
@@ -269,7 +271,8 @@ struct uk_event {
+enum uk_event_status uk_raise_event_ptr(struct uk_event *e, void *data)
 {
 	const uk_event_handler_t *itr;
 	int rc;
```
Shouldn't `rc` and `ret` also be enum given the comparisons below?

michpappas replied:
Yes, and that makes me realize that I completely missed the case that the handlers can also return errors. Which means that `enum` won't do. I'll push an update to address the problem in a different way.

mogasergiu commented on lib/ukdebug/arch/arm64/gdbsup.c:
```diff
@@ -79,7 +79,7 @@ static int gdb_arch_check_brk(unsigned long pc)
 }
 
-static int gdb_arch_debug_handler(void *data)
+static enum uk_event_status gdb_arch_debug_handler(void *data)
```
`have_brk` and `r` are `int`

mogasergiu commented on lib/ukdebug/arch/x86_64/gdbsup.c:
```diff
@@ -37,7 +37,7 @@ static int gdb_arch_dbg_trap(int errnr, struct __regs *regs)
 }
 
-static int gdb_arch_debug_handler(void *data)
+static enum uk_event_status gdb_arch_debug_handler(void *data)
```
Ditto `r`.

mogasergiu commented on lib/uknofault/excpttab.h:
```diff
@@ -78,7 +78,8 @@ extern const struct nf_excpttab_entry uk_excpttab_end[];
 
-static inline int nf_handle_trap(__vaddr_t ip, struct ukarch_trap_ctx *ctx)
+static inline
+enum uk_event_status nf_handle_trap(__vaddr_t ip, struct ukarch_trap_ctx *ctx)
```
There is a `return handler` statement and it seems that the definition is to return int.
```C
typedef int (*nf_excpt_handler)(const struct nf_excpttab_entry *e,
				struct ukarch_trap_ctx *ctx);
```

mogasergiu reviewed (changes requested):
LGTM, just a couple of missed type updated I think

michpappas committed:
plat/xen: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/ukintctlr: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
lib/posix-process: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

michpappas committed:
drivers/ukps2/kbd: Adapt to strongly typed uk_event_status API

Update event handlers and event raising calls to use the new
enum return type and error parameter.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on include/uk/event.h:
```diff
@@ -74,33 +74,34 @@
- *       else if (!rc)
- *               uk_pr_err("myevent not handled!\n");
+ *       rc = uk_raise_event(myevent, &myevent_data, &error);
+ *       if (unlikely(rc < UK_EVENT_ERROR))
```
I would drop the unlikely from the documentation and leave as it was. Also the type of `rc` might need to be updated.

mogasergiu commented on include/uk/event.h:
```diff
@@ -259,32 +268,35 @@ struct uk_event {
 {
 	const uk_event_handler_t *itr;
-	int rc;
 	int ret = UK_EVENT_NOT_HANDLED;
```
I think `ret` should have been enum?

mogasergiu commented on include/uk/event.h:
```diff
@@ -259,32 +268,35 @@ struct uk_event {
 			return UK_EVENT_HANDLED;
 
-		if (rc == UK_EVENT_HANDLED_CONT)
+		if (ret == UK_EVENT_HANDLED_CONT)
```
I think this if statement is no longer necessary

mogasergiu commented on include/uk/event.h:
```diff
@@ -302,9 +314,9 @@ static inline int uk_raise_event_ptr(struct uk_event *e, void *data)
+#define uk_raise_event(event, data, error)				\
 	({	_UK_EVT_IMPORT_EVENT(event);				\
-		uk_raise_event_ptr(UK_EVENT_PTR(event), data);	})
+		uk_raise_event_ptr(UK_EVENT_PTR(event), data, error);	})
```
Maybe we should enclose `data` and `error` between paranthesis? Also don't forget to now add the `error` argument documentation to `uk_raise_event` as well.

mogasergiu commented on lib/posix-process/execve.c:
```diff
@@ -83,9 +83,12 @@ static int pprocess_cleanup(struct posix_process *pprocess)
-int pprocess_raise_execve_event(struct posix_process_execve_event_data *data)
+enum uk_event_status
+pprocess_raise_execve_event(struct posix_process_execve_event_data *event_data,
+			    int event_error)
```
I think you meant `int *event_error`? (You do have it in the header definition :D)

mogasergiu commented on lib/ukboot/shutdown_req.c:
```diff
@@ -57,7 +58,8 @@ int uk_boot_shutdown_req_isr(enum ukplat_gstate target)
 #if __INTERRUPTSAFE__ && CONFIG_LIBUKBOOT_SHUTDOWNREQ_HANDLER
-static int shutdown_req_handler(void *data)
+static
+enum uk_event_status shutdown_req_handler(void *data, int *error __unused)
```
I think you can update `error` to return value of `rc`.

mogasergiu commented on lib/ukvmem/arch/arm/pagefault64.c:
```diff
@@ -6,14 +6,16 @@
+#include <uk/print.h>
 
-static int vmem_arch_pagefault(void *data)
+static enum uk_event_status vmem_arch_pagefault(void *data, int *error __unused)
```
Ditto (for x86 too)

mogasergiu reviewed (changes requested):
For me it feels viscerally wrong whenever I end up having to add functions with lateral effect and can't help but wonder if there is a better way. But, yeah, for now I can't see a better way than this that would fit all requirements either...
So I guess we can stick with this or... maybe leave it for later and brainstorm some more. I would approve either way.
